Prep libjpeg-turbo for runtimecore

colingpowell opened the pull request:
Issue: https://devtopia.esri.com/runtime/home-one/issues/470

Update `libjpeg-turbo` build settings to make it usable in Runtime Core.

- [x] Library builds for all platforms: 
  - [x86 ](http://runtime-test.esri.com:8080/view/DEV/job/dev-vtest-3rdparty_libraries-interface/147/downstreambuildview/)
  - [x64](http://runtime-test.esri.com:8080/view/DEV/job/dev-vtest-3rdparty_libraries-interface/148/downstreambuildview/)
  - [arm](http://runtime-test.esri.com:8080/view/DEV/job/dev-vtest-3rdparty_libraries-interface/150/downstreambuildview/)
  - [arm64](http://runtime-test.esri.com:8080/view/DEV/job/dev-vtest-3rdparty_libraries-interface/150/downstreambuildview/)
- [x] runtimecore + tests build with this library installed (a few baselines will need to be updated): 
  - [x86 ](http://runtime-test.esri.com:8080/view/vTest/job/vtest_interface/3009/)
  - [x64](http://runtime-test.esri.com:8080/view/vTest/job/vtest_interface/3008/downstreambuildview/)
  - [arm](http://runtime-test.esri.com:8080/view/vTest/job/vtest_interface/3010/downstreambuildview/)
  - [arm64](http://runtime-test.esri.com:8080/view/vTest/job/vtest_interface/3007/downstreambuildview/)

**Build Config**
Config settings are located in `jconfig.h` and `jconfigint.h`, which are based on the `libjpeg-turbo` config files located in the `skia` library.

`cmake` was run locally to auto-generate these config files and compare them those ported from `skia`. The `cmake`-generated config files included the pair of `#if (defined _WIN32)` blocks that were added to our own `jconfig.h` in 8206c5c. 

**SIMD Support**
SIMD support (the `turbo` in `libjpeg-turbo`) has been added to `arm` and `arm64`-equipped iOS and Android devices. `libjpeg-turbo` requires the `NASM` assembler for SIMD support on `x86` and `x64` architectures, and it seemed best to avoid introducing another binary into our 3rd party ecosystem solely to assemble native SIMD commands for this library. 

SIMD support is not included for WinUWP platforms, as VS projects don't presently allow files to be conditionally included in a project based on its platform.

For platforms that support SIMD, source files in `simd/[architecture]` are compiled, which contain SIMD-specific versions of various JPEG compression and decompression algorithms. Otherwise, `jsimd_none.c` is compiled, which defines these SIMD-specific functions as empty stubs.

All of this allows for the correct compression/decompression functions to be selected at runtime - for example:
```
#ifdef DCT_IFAST_SUPPORTED
      case JDCT_IFAST:
        if (jsimd_can_idct_ifast())
          method_ptr = jsimd_idct_ifast;
        else
          method_ptr = jpeg_idct_ifast;
        method = JDCT_IFAST;
        break;
#endif
```
The `jsimd_can_` functions are defined in both `jsimd_none.c` and the `simd/[architecture]` source files. They work essentially like flags - for example: 
```
GLOBAL(int)
jsimd_can_idct_ifast(void)
{
  return 0;
}
```
Where this particular functionality is allowed, this function is defined to return 1. 

Our build system appears to disallow building from identically-named source files located in different directories. Unfortunately, the source files located in each `/simd/[architecture]` directory follow this pattern. To work-around this limitation, we created a pair of uniquely-named source files in each directory and simply `#include`d the offending files there.  

Please review: @chri7325 @john4744 @dkalowsk

itrombley committed:
Add for runtimecore

itrombley committed:
Add for runtimecore

itrombley committed:
Rename lua file

colingpowell committed:
libjpeg-turbo: add SIMD support for arm devices

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: specify jsimd_none.c

Build with jsimd_none.c only on platforms that do not support the SIMD
functions bundled with this library.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: add _CRT_SECURE_NO_WARNINGS

Add _CRT_SECURE_NO_WARNINGS flag to WinUWP builds, as otherwise MSVC will
throw errors stemming from the use of sscanf and sprintf.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: workaround build sys limits for arm

Create a new file in simd/arm corresponding to each existing file,
give the new files unique names, and directly include each existing
file in the matching new file.

This circumvents limitations in our build system, which doesn't
support identically-named files located in different directories.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: wkaround build sys limits for arm64

Continue work started in previous commit, making similar changes for
arm64 platforms.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: update config header

Update jconfig.h to prevent jmorecfg.h from redefining boolean and
int32 types on Windows platforms.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: update version

Update libjpeg-turbo version - matching what is being built from this
branch.

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: remove extraneous lines in lua

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: update lua UUID

Signed-off by: Colin Powell <cpowell@esri.com>

john4744 reviewed (approved):
looks good 👍

chri7325 commented on .gitignore:
```diff
@@ -0,0 +1 @@
+Makefile
```
Can this be the RTC one so it'll help with all platforms? https://devtopia.esri.com/3rdparty/zlib/blob/runtimecore/.gitignore

chri7325 commented on libjpeg-turbo.lua:
```diff
@@ -0,0 +1,606 @@
+
+  if (os.is("windows")) then
+
+    defines {
```
Can the defins and files move to the appropriate windows blocks?

chri7325 commented on libjpeg-turbo.lua:
```diff
@@ -0,0 +1,606 @@
+    configuration { "ios*" }
+
+      defines {
+        "_REENTRANT",
```
Are these defines needed?

colingpowell replied:
Looks like these (and the android defines) were brought over from the old `jpeg.lua`...I'll remove them and confirm that everything still builds.

chri7325 commented on libjpeg-turbo.lua:
```diff
@@ -0,0 +1,606 @@
+    configuration { "android*" }
+
+      defines {
+        "HAVE_MEMMOVE",
```
Is this define needed?

chri7325 commented on libjpeg-turbo.lua:
```diff
@@ -0,0 +1,606 @@
+
+    dofile (_BUILD_DIR .. "/static_winuwp.lua")
+
+    defines {
```
Can this be tabbed over? It should also go after the configuration below

chri7325 commented on libjpeg-turbo.lua:
```diff
@@ -0,0 +1,606 @@
+    }
+
+    files {
+      "jsimd_none.c",
```
Can the jsimd files be put into a table above so they can easily be updated in the future.  Like https://devtopia.esri.com/3rdparty/skia/blob/runtimecore/skia.lua#L344

chri7325 reviewed (approved):
Looks great!  Just a few minor suggestions.

chri7325 commented:
Oh, same request for adding the cmake command that was used to generate the files in the commit that brought them in.

colingpowell committed:
fixup: address premake script feedback

- Move jsimd files into tables and reference each table where needed
- Remove unnecessary defines
- Move Windows defines / build files to appropriate sections of script

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: use RTC-standard .gitignore

Signed-off by: Colin Powell <cpowell@esri.com>

colingpowell committed:
libjpeg-turbo: combine build configs

Create a single jconfig.h and jconfigint.h that are
the union of all versions of these files generated
for all supported platforms. Files were generated
using the cmake scripts documented in BUILDING.MD.

Config files for Windows platforms differed slightly than those
produced for Unix-y platforms (iOS, MacOS, Linux, Android);
`#if(defined _WIN32)` blocks were used where necessary
to specify Windows-only settings.

The value of SIZE_OF_SIZE_T differed between 32-bit and 64-bit
build configs; the #ifdef blocks used here to set this value
appropriately are adapted from the libjpeg-turbo config files
located in the skia library.

Signed-off by: Colin Powell <cpowell@esri.com>

chri7325 reviewed (approved):
Looks great!
